### PR TITLE
(GH-2522) Add 'resolve' key to modules config

### PR DESCRIPTION
**Pull request by beechtom**
This adds support for a new `resolve` key for both Forge and git module
specs in `bolt-project.yaml`. It also adds support for a `name` key for
git module specs, which is required when setting the `resolve` key to
`false`.

When `resolve` is `false`, Bolt will skip dependency resolution for the
module. This allows users to include modules that would otherwise break
Bolt's dependency resolution in their Bolt projects, avoiding the need
to manually manage a Puppetfile. This includes modules with broken
metadata and git modules hosted in a repository other than a public
GitHub repository.

If a unresolved module shares the name of another module's dependency,
Bolt will error indicating that there is a name conflict. This is in
line with r10k's behavior when installing a Puppetfile that has
duplicate module names.

!feature

* **Add `resolve` key for Forge and git module specifications**
  ([#2522](https://github.com/puppetlabs/bolt/issues/2522))

  Forge and git module specifications in `bolt-project.yaml` now support
  a `resolve` key. When setting `resolve: false`, Bolt will skip
  dependency resolution for the module, allowing users to include
  modules with broken metadata or modules hosted in a repository other
  than a public GitHub repository in their project configuration.

---

### Test cases

- [ ] Requires git modules to have a `name` if `resolve: false`
- [ ] Skips resolution for modules with `resolve: false` when installing and adding modules
- [ ] Allows non-public or non-GitHub git modules to be installed without `--no-resolve`
- [ ] Allows modules with broken metadata to be installed
- [ ] Errors if no-resolve modules have a name conflict with resolved dependencies

### CODE REVIEW

**Comment by lucywyman** [comment]
Woops, quick rebase.

File: lib/bolt/module_installer.rb
Comment (lucywyman) [inline lib/bolt/module_installer.rb:198]: When is this false?

Comment (beechtom) [reply lib/bolt/module_installer.rb:198]: Anytime installing modules fails. I specifically ran into this when a Puppetfile has duplicate module names -- r10k will error and fail to install modules, but Bolt will continue to try and generate types.

Comment (lucywyman) [reply lib/bolt/module_installer.rb:198]: It seems like we should raise an error here if `ok` is false in that case.

File: lib/bolt/config/options.rb
Comment (lucywyman) [inline lib/bolt/config/options.rb:265]: Hm, this weirdly doesn't seem to error with non-boolean values are provided. Do we not run the validator when doing `module install`?

Comment (beechtom) [reply lib/bolt/config/options.rb:265]: No, the validator would need to be updated to handle the `oneOf` key since there are multiple valid definitions for a module spec. I can add in a separate check as part of the module objects to ensure that it's a boolean.

File: lib/bolt/module_installer.rb
Comment (beechtom) [reply lib/bolt/module_installer.rb:198]: My only hesitation with doing that is we explicitly have a check in the outputter to print a message stating that installing the modules failed https://github.com/puppetlabs/bolt/blob/main/lib/bolt/outputter/human.rb#L428-L434 . I'm fine with making the change, though, if we want to error instead.

File: lib/bolt/config/options.rb
Comment (lucywyman) [reply lib/bolt/config/options.rb:265]: That'd be great.

Comment (lucywyman) [reply lib/bolt/config/options.rb:265]: Actually, it seems like the validator should be checking every key that's there against it's type even if there are multiple possible configurations (basically should be able to handle `oneOf`). It looks like this is the only instance of `oneOf`, so it might be appropriate to add here.

File: documentation/bolt_installing_modules.md
Comment (lucywyman) [inline documentation/bolt_installing_modules.md:371]: I think it's also worth calling out (here or elsewhere) that the value of `name` is how the module will be named on the modulepath as well.

Comment (beechtom) [reply documentation/bolt_installing_modules.md:371]: Added some more info to the table further up the page that lists the available keys.

File: spec/bolt/module_installer/specs/git_spec_spec.rb
Comment (lucywyman) [inline spec/bolt/module_installer/specs/git_spec_spec.rb:48]: Since it's a separate check it'd also be good to verify that `resolve: <non-boolean>` errors for both forge and git modules.

**Review by lucywyman — CHANGES_REQUESTED** [review]

File: documentation/bolt_installing_modules.md
Comment (adreyer) [inline documentation/bolt_installing_modules.md:218]: This should be more explicit about whether it's namespaced or not.

File: documentation/bolt_installing_modules.md
Comment (adreyer) [inline documentation/bolt_installing_modules.md:345]: + The modules dependencies cannot be cleanly resolved

File: documentation/bolt_installing_modules.md
Comment (beechtom) [reply documentation/bolt_installing_modules.md]: Updated. Not 100% sure on the phrasing, though... for example, if the name doesn't match the module's _actual_ name, tasks can still be run but not plans.

**Comment by lucywyman** [comment]
- [x] Does not install dependencies when `resolve: false`
- [x] Installs dependencies when `resolve` is:
  - [x] Unset
  - [x] True
- [x] Errors if resolve isn't boolean
- [x] Errors if resolve is set for a git module and name isn't set
- [x] Git module with a name different from actual name uses provided name
- [x] Adding a fake module with `resolve: false` doesn't fail until installing modules
  - [x] Adding a fake module with `resolve: false` doesn't prevent other modules from being installed
- [x] Errors when `resolve: false` module has the same name as a module dependency

**Review by lucywyman — APPROVED** [review]
Just needs docs review.

File: documentation/bolt_installing_modules.md
Comment (hestonhoffman) [inline documentation/bolt_installing_modules.md:188]:
```suggestion
| `resolve` | Boolean. Whether to resolve the module's dependencies when installing modules. | |
```

File: documentation/bolt_installing_modules.md
Comment (hestonhoffman) [inline documentation/bolt_installing_modules.md:218]:
```suggestion
| `name` | The name of the module. Bolt uses this name for the module in the Puppetfile, the directory that the module's contents are downloaded to, and as a namespace for the module's content. To avoid errors, make sure this name matches the name specified in the module's `metadata.json`. **Required if `resolve` is `false`.** | |
```

File: documentation/bolt_installing_modules.md
Comment (hestonhoffman) [inline documentation/bolt_installing_modules.md:220]:
```suggestion
| `resolve` | Boolean. Whether to resolve the module's dependencies when installing modules. | |
```

File: documentation/bolt_installing_modules.md
Comment (hestonhoffman) [inline documentation/bolt_installing_modules.md:341]:
```suggestion
Skipping dependency resolution for a module allows you to
take advantage of Bolt's module dependency resolution while still using modules
that Bolt cannot resolve dependencies for.

You might want to skip dependency resolution for a module if:
```
Just streamlining this a bit because it's a little redundant with the list

File: documentation/bolt_installing_modules.md
Comment (hestonhoffman) [inline documentation/bolt_installing_modules.md:346]:
```suggestion
- Bolt can't cleanly resolve the module's dependencies.
```

**Review by lucywyman — APPROVED** [review]
Will merge once Heston approves as well (no rush Heston!) :sparkle:

**Review by hestonhoffman — APPROVED** [review]
Docs LGTM!

**Pull request merged**
